### PR TITLE
feat(action-harness): Production in-process Flashblocks Builder

### PR DESCRIPTION
**Pull request by refcell**
Adds a production-builder backend to the action test harness so action tests can exercise real Base block-construction policy -- pool-based transaction selection, bundle validity windows, and DA/gas metering as coverage grows -- which the default in-memory engine never reaches since it builds with a noop pool.

BuilderBackedEngineClient launches an in-process LocalInstance running the production FlashblocksServiceBuilder against the harness's rollup-derived genesis and drives it over the real Engine API through the production SequencerEngineClient seam. L2Sequencer is now generic over the engine backend, defaulting to the existing ActionEngineClient so all current action tests are unchanged, and harness-injected transactions route through the real mempool instead of being force-included.

New tests in actions/harness/tests/real_builder.rs cover block production through the actor, pool-based transaction selection, bundle validity-window enforcement, and a full builder-to-batcher-to-derivation round trip verifying state-root parity with the verifier.

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>✅ Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
✅
<code>1/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: actions/harness/src/sequencer/builder_backend.rs [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+
+    async fn start_build_block(
+        &self,
+        attributes: AttributesWithParent,
+    ) -> EngineClientResult<PayloadId> {
+        let parent = attributes.parent.block_info.hash;
+        let builder_attrs = BasePayloadBuilderAttributes::<BaseTxEnvelope>::try_new(
+            parent,
+            attributes.attributes,
+            3,
```
[the remark this answers is not in the document]
Comment (refcell) [reply actions/harness/src/sequencer/builder_backend.rs:181]: Investigated this coupling. `BasePayloadBuilderAttributes` currently recomputes payload IDs with version 3 in its `PayloadAttributes` implementation (`crates/execution/payload/src/payload.rs`), so changing only this constructor to 4 would make `getPayload` miss the engine-registered build. I kept the required version 3 and documented the engine-side contract at the call site. Making payload IDs version-aware needs to be fixed at the shared payload-attributes source of truth rather than patched only in this backend.

File: actions/harness/src/sequencer/builder_backend.rs [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+            .await
+            .map_err(|e| EngineClientError::ResponseError(e.to_string()))?;
+        Ok(BaseExecutionPayloadEnvelope {
+            parent_beacon_block_root: attributes
+                .attributes
+                .payload_attributes
+                .parent_beacon_block_root,
+            execution_payload: BaseExecutionPayload::V4(envelope.execution_payload),
+        })
+    }
```
[the remark this answers is not in the document]
Comment (refcell) [reply actions/harness/src/sequencer/builder_backend.rs:247]: Fixed in af254beac. The backend now retains each sealed payload’s EIP-7685 requests keyed by block hash and supplies those exact requests to `newPayload`; it errors instead of silently substituting defaults when metadata is missing.

File: actions/harness/src/sequencer/builder_backend.rs [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        let pooled = BasePooledTransaction::new(recovered, encoded_length).with_bundle_metadata(
+            min_block,
+            max_block,
+            min_timestamp_millis,
+            max_timestamp_millis,
+        );
+        pool.add_external_transaction(pooled)
+            .await
+            .map_err(|e| eyre::eyre!("pool rejected bundle transaction: {e}"))?;
+        Ok(())
```
[the remark this answers is not in the document]
Comment (refcell) [reply actions/harness/src/sequencer/builder_backend.rs:138]: Fixed in af254beac. `inject_bundle_transaction` now returns `EngineClientResult<()>` and maps signer recovery and pool rejection through `EngineClientError::RequestError`, matching regular pool injection.

File: actions/harness/src/sequencer/builder_backend.rs [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+            return Err(EngineClientError::RequestError(format!(
+                "engine rejected payload: {status:?}"
+            )));
+        }
+
+        let current_head = self.head.lock().expect("head lock").block_info.hash;
+        engine
+            .update_forkchoice(current_head, new_hash, None)
+            .await
+            .map_err(|e| EngineClientError::RequestError(e.to_string()))?;
```
[the remark this answers is not in the document]
Comment (refcell) [reply actions/harness/src/sequencer/builder_backend.rs:283]: Fixed in af254beac. The canonical forkchoice response is now required to be VALID before requests are cleared or the harness registry/head is advanced; non-VALID responses propagate as engine errors.

File: actions/harness/src/sequencer/builder_backend.rs [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+                .try_into_recovered()
+                .map_err(|e| EngineClientError::RequestError(format!("recover signer: {e}")))?;
+            let encoded_length = recovered.encode_2718_len();
+            let pooled = BasePooledTransaction::new(recovered, encoded_length);
+            pool.add_external_transaction(pooled)
+                .await
+                .map_err(|e| EngineClientError::RequestError(e.to_string()))?;
+        }
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (refcell) [reply actions/harness/src/sequencer/builder_backend.rs:336]: Addressed in a0122ad8e. The external pool handle only supports single-transaction insertion and cannot provide atomic batch semantics or rollback, so the `SequencerEngineBackend` contract now explicitly documents that an error may leave an accepted prefix queued for a later block. This makes the required caller behavior clear without claiming atomicity the pool API cannot guarantee.

File: actions/harness/src/sequencer/builder_backend.rs [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+            .pending_execution_requests
+            .lock()
+            .expect("pending execution requests lock")
+            .get(&new_hash)
+            .cloned()
+            .ok_or_else(|| {
+                EngineClientError::RequestError(format!(
+                    "missing execution requests for payload {new_hash}"
+                ))
+            })?;
```
[the remark this answers is not in the document]
Comment (refcell) [reply actions/harness/src/sequencer/builder_backend.rs]: Fixed in a0122ad8e. `insert_unsafe_payload` now removes the request metadata before making engine calls, so INVALID payloads, rejected FCUs, and transport errors cannot leak entries.

File: actions/harness/src/sequencer/builder_backend.rs [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+            3,
+        )
+        .map_err(|e| EngineClientError::RequestError(e.to_string()))?;
+        let fcu = self
+            .engine()
+            .update_forkchoice(parent, parent, Some(builder_attrs))
+            .await
+            .map_err(|e| EngineClientError::RequestError(e.to_string()))?;
+        fcu.payload_id
+            .ok_or_else(|| EngineClientError::ResponseError("no payload id returned".into()))
```
[the remark this answers is not in the document]
Comment (refcell) [reply actions/harness/src/sequencer/builder_backend.rs:196]: Fixed in ec7eefd53. `start_build_block` now rejects non-VALID forkchoice responses before reading `payload_id`, and classifies the engine-returned rejection as `ResponseError` so callers see the actual FCU failure instead of a later unknown-payload error.

File: actions/harness/src/batcher/tx_manager.rs [hunk 8]
```diff
@@ -141,6 +143,23 @@ impl L1MinerTxManager {
...
     }
 
+    /// Wait until the driver has queued at least one pending submission.
+    ///
+    /// Unlike yielding to the scheduler, this provides a deterministic handoff
+    /// between the background [`BatchDriver`] and the harness before an L1 block
+    /// is mined.
+    ///
+    /// [`BatchDriver`]: base_batcher_core::BatchDriver
+    pub async fn wait_for_pending(&self) {
```
[the remark this answers is not in the document]
Comment (refcell) [reply actions/harness/src/batcher/tx_manager.rs]: Fixed in ec7eefd53. The pending-submission handoff is now bounded by the driver’s 10-second timeout and returns `BatcherError::SubmissionTimeout` instead of hanging. I also made new cycles reject outstanding pending/staged submissions, preventing stale work from satisfying the wait, and enabled Tokio’s `time` feature directly.

File: actions/harness/src/sequencer/builder_backend.rs [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+        let block = ExecutionPayloadConverter::block_from_envelope(&payload)
+            .map_err(|e| EngineClientError::RequestError(e.to_string()))?;
+        let new_hash = block.header.hash_slow();
+        let parent_beacon_block_root = payload.parent_beacon_block_root.unwrap_or_default();
+        let BaseExecutionPayload::V4(v4) = payload.execution_payload else {
+            return Err(EngineClientError::RequestError(
+                "builder backend expects a V4 execution payload".into(),
+            ));
+        };
+        let execution_requests = self
```
Comment (xenoliss) [inline actions/harness/src/sequencer/builder_backend.rs]: Base currently rejects non-empty execution requests and reconstructs V4 payloads with EMPTY_REQUESTS_HASH, so this side map adds non-idempotent state without real support.
Could we assert requests are empty and pass `Requests::default()` to `newPayload`, removing the map entirely?

File: actions/harness/src/batcher/actor.rs [hunk 10]
```diff
@@ -241,7 +251,9 @@ impl<S: L2BlockProvider> Batcher<S> {
             return Err(BatcherError::NoBlocks);
         }
         self.block_tx.send(L2BlockEvent::Flush).expect("driver task alive");
-        tokio::task::yield_now().await;
+        if !self.tx_manager.wait_for_pending(Duration::from_secs(10)).await {
```
Comment (xenoliss) [inline actions/harness/src/batcher/actor.rs]: `wait_for_pending` returns after the first queued frame, while `advance()` promises to queue all frame submissions before mining. Later frames can miss the L1 block on a multi-thread runtime.
Could we wait for an explicit flush-complete acknowledgement?

File: actions/harness/src/sequencer/builder_backend.rs [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+            .map_err(|e| EngineClientError::RequestError(e.to_string()))?;
+        if !status.is_valid() {
+            return Err(EngineClientError::RequestError(format!(
+                "engine rejected payload: {status:?}"
+            )));
+        }
+
+        let current_head = self.head.lock().expect("head lock").block_info.hash;
+        let fcu = engine
+            .update_forkchoice(current_head, new_hash, None)
```
Comment (xenoliss) [inline actions/harness/src/sequencer/builder_backend.rs:281]: This helper promotes `current_head` to both safe and finalized, so the backend cannot model unsafe reorgs. Worth documenting this as a known limitation until the planned P4 forkchoice work?

File: actions/harness/src/sequencer/builder_backend.rs [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+                .payload_attributes
+                .parent_beacon_block_root,
+            execution_payload: BaseExecutionPayload::V4(envelope.execution_payload),
+        };
+        let block = ExecutionPayloadConverter::block_from_envelope(&payload)
+            .map_err(|e| EngineClientError::ResponseError(e.to_string()))?;
+        self.pending_execution_requests
+            .lock()
+            .expect("pending execution requests lock")
+            .insert(block.header.hash_slow(), execution_requests);
```
[the remark this answers is not in the document]
Comment (refcell) [reply actions/harness/src/sequencer/builder_backend.rs]: Fixed in 9695608. `get_sealed_payload` no longer computes `block_from_envelope`/`hash_slow()` at all — the whole `pending_execution_requests` map is gone (see the reply on the sibling thread), so there's nothing to key by hash anymore. `insert_unsafe_payload` still computes the hash once, but only for the purpose it always needed it for (the FCU new-head argument), not for a map lookup.

File: actions/harness/src/sequencer/builder_backend.rs (hunk 9 above)
Comment (refcell) [reply actions/harness/src/sequencer/builder_backend.rs]: Fixed in 9695608. Confirmed the underlying claim: the production builder hardcodes `execution_requests: vec![]` when sealing payloads (`crates/execution/payload/src/payload.rs`), and post-Isthmus header validation independently rejects any non-empty `requests_hash` — so the map genuinely couldn't hold anything but empty requests. Removed `pending_execution_requests` entirely; `get_sealed_payload` now asserts `envelope.execution_requests.is_empty()` (erroring otherwise) and `insert_unsafe_payload` passes `Requests::default()` straight through.

File: actions/harness/src/sequencer/builder_backend.rs (hunk 11 above)
Comment (refcell) [reply actions/harness/src/sequencer/builder_backend.rs:281]: Documented in 9695608 as a known limitation on `insert_unsafe_payload`, pointing at the planned P4 forkchoice work as you suggested.

File: actions/harness/src/batcher/actor.rs (hunk 10 above)
Comment (refcell) [reply actions/harness/src/batcher/actor.rs]: Good catch — this was real, not just theoretical: `real_builder.rs` is the first caller to run the batcher under a multi-thread runtime, which is what turns "wakes on the first pending item" into an actual race between the notifying driver thread and the waiting harness thread. Fixed in 066938298 with a real settlement signal rather than a heuristic: `L2BlockEvent::Flush` (and `AdminCommand::Flush`) now carry an optional ack that `BatchDriver::run` only fires once a later CPU phase reports *both* encoding and submission fully drained for the channel — i.e. every frame from the flush, not just the first, has been handed to the tx manager. The ack rides the same ordered channel as the preceding `Block` events (not the admin side-channel) so it can't observe stale state. Added two regression tests in `batcher/core/src/driver.rs` covering the multi-frame settlement and the backpressure (ack must not fire while a ready submission is still waiting on semaphore capacity) cases.

File: crates/batcher/source/src/event.rs [hunk 13]
```diff
@@ -2,9 +2,10 @@
 
 use base_common_consensus::BaseBlock;
 use base_protocol::L2BlockInfo;
+use tokio::sync::oneshot;
 
 /// Events emitted by an [`UnsafeBlockSource`][crate::UnsafeBlockSource].
-#[derive(Debug, Clone)]
+#[derive(Debug)]
```
[the remark this answers is not in the document]
Comment (refcell) [reply crates/batcher/source/src/event.rs:14]: Fair point, though worth noting: this crate is unpublished (workspace version 0.0.0, path-only within this monorepo), so there's no external semver contract to break — a full `cargo check --workspace` (which I ran) is the actual compatibility bar here, and it passes clean since nothing clones `L2BlockEvent` today. Went with documenting the trade-off in 3043b028 rather than the `Arc<Mutex<Option<_>>>` wrapper: that would let the same ack fire twice (once per clone) or be silently dropped, which is worse than just not being `Clone`. Added a doc note on the enum explaining why, so it doesn't get re-added by accident later.

File: crates/batcher/core/src/driver.rs [hunk 14]
```diff
@@ -156,13 +164,23 @@ where
...
+
+            if encoding_drained && submissions_drained && !self.pending_flush_acks.is_empty() {
+                debug!(
+                    acks = %self.pending_flush_acks.len(),
+                    "flush settled: encoding and submission fully drained"
+                );
+                for ack in self.pending_flush_acks.drain(..) {
+                    let _ = ack.send(());
+                }
+            }
```
[the remark this answers is not in the document]
Comment (refcell) [reply crates/batcher/core/src/driver.rs:191]: Confirmed and agreed — this is a real precision gap, not just a nitpick: the ack is a sufficient-but-not-tight signal (it can never fire before this flush's own frames are submitted, but concurrent `Block` events arriving afterward can delay it, and under sustained continuous ingestion it may never fire at all).

I considered the per-flush-watermark approach (tagging submissions with the generation that triggered them so the ack could fire precisely), but that requires plumbing submission-generation tracking through `BatchPipeline`/`BatchEncoder" in `base-batcher-encoder` — a materially bigger change than this PR's scope (adding a settlement signal for the harness's flush), and not needed by any current caller: the action-test harness is the only caller today and it always quiesces (guarded by `OutstandingSubmissions`) before flushing, so it gets a precise, always-terminating signal in practice.

Went with your first suggested option in 3043b028: documented the exact scope/caveat on `AdminCommand::Flush`, `AdminHandle::flush_and_wait`, `L2BlockEvent::Flush`, and `BatchDriver::run`, so a future production caller streaming blocks concurrently knows the guarantee is weaker than it might assume. Happy to revisit with the watermark approach if/when a caller actually needs it under continuous load.

File: crates/batcher/core/src/driver.rs [hunk 15]
```diff
@@ -353,11 +396,11 @@ where
                 }
 
                 event = self.source.next() => match event {
-                    Ok(L2BlockEvent::Block(_) | L2BlockEvent::Flush) if self.stopped => {
+                    Ok(L2BlockEvent::Block(_) | L2BlockEvent::Flush { .. }) if self.stopped => {
                         continue;
```
[the remark this answers is not in the document]
Comment (refcell) [reply crates/batcher/core/src/driver.rs:410]: Agreed this deserved better signal — fixed in aab623d2. Split the combined `Block | Flush` "stopped" arm so the `Flush` ack is handled explicitly: still dropped (a paused batcher produces no frames, so firing would falsely report settlement), but now with a `debug!` log explaining why, rather than an unannotated `continue`. To be clear on the actual behavior though: this was never a hang — dropping a oneshot sender resolves the receiver's `.await` immediately with `RecvError`, it just didn't say why. Also extended `test_paused_drops_block_and_flush_events` to actually send a `Flush{ack}` while paused and assert the receiver errors, since the existing test's doc comment claimed to cover both `Block` and `Flush` but only exercised `Block`.

File: crates/batcher/core/src/submissions.rs [hunk 16]
```diff
@@ -85,18 +85,22 @@ impl<TM: TxManager> SubmissionQueue<TM> {
...
+    /// txpool is blocked, or a blob-encoding failure required a requeue.
+    pub async fn submit_pending<P: BatchPipeline>(&mut self, pipeline: &mut P) -> bool {
         loop {
             if self.txpool_blocked {
-                break;
+                return false;
             }
             let Ok(permit) = Arc::clone(&self.semaphore).try_acquire_owned() else {
-                break;
+                return false;
```
Comment (0x00101010) [inline crates/batcher/core/src/submissions.rs]: If exactly max_pending submissions are queued, all are handed to TxManager, but this returns false before noticing the pipeline is empty. The flush ack then waits for a receipt while the harness waits for the ack before mining, causing a timeout. Could we check whether work remains before reporting backpressure?

File: actions/harness/src/sequencer/builder_backend.rs [hunk 17]
```diff
@@ -0,0 +1,416 @@
...
+        &self,
+        payload_id: PayloadId,
+        attributes: AttributesWithParent,
+    ) -> EngineClientResult<BaseExecutionPayloadEnvelope> {
+        // Give the flashblocks build loop a full block time to assemble the block before resolving
+        // it, matching the production sequencer's start-of-slot to end-of-slot cadence.
+        tokio::time::sleep(self.block_time).await;
+        let envelope = self
+            .engine()
+            .get_payload(payload_id)
```
Comment (0x00101010) [inline actions/harness/src/sequencer/builder_backend.rs]: This always calls `engine_getPayloadV4`, but Azul requires getPayloadV5 (while still importing with newPayloadV4). Could we select the retrieval version from the payload timestamp so builder-backed Azul/Beryl/Cobalt tests work?

File: crates/batcher/core/src/submissions.rs (hunk 16 above)
Comment (refcell) [reply crates/batcher/core/src/submissions.rs]: Good catch — this was a real false-negative, not just a corner case. Fixed in 81f458a64: added `BatchPipeline::has_ready_submission()`, a non-mutating peek. When the semaphore is exhausted, `submit_pending` now checks that instead of unconditionally reporting backpressure — if the pipeline actually has no more ready work (e.g. exactly `max_pending` submissions were queued and all handed to the tx manager), it reports fully drained so the flush ack can fire; if a ready submission is genuinely still waiting on capacity, it still reports backpressure exactly as before.

Added two unit tests in `crates/batcher/core/src/submissions.rs` covering both cases (exact-fill → drained, more-than-capacity → backpressured), and verified none of the existing semaphore/backpressure driver tests regressed.

File: actions/harness/src/sequencer/builder_backend.rs (hunk 17 above)
Comment (refcell) [reply actions/harness/src/sequencer/builder_backend.rs]: Fixed in 81f458a64. `get_sealed_payload` now selects the retrieval version from the payload timestamp via `EngineGetPayloadVersion::from_cfg(&self.rollup_config, timestamp)`: `engine_getPayloadV5` once Base Azul is active, `engine_getPayloadV4` otherwise — mirroring the production `EngineTaskQueue::fetch_payload` version selection. `newPayloadV4` in `insert_unsafe_payload` is left as-is per your note, since both V4 and V5 envelopes carry the same `BaseExecutionPayloadV4` payload shape.

Added `EngineApi::get_payload_v5` to `base-builder-core`'s test-utils Engine API client, and a regression test (`builder_backed_sequencer_produces_block_with_azul_active` in `actions/harness/tests/real_builder.rs`) that activates Azul at genesis and drives a block through the real builder. Verified it fails against the pre-fix code with `ErrorObject { code: ServerError(-38005), message: "Unsupported fork" }`, and passes with the fix.

**Review by 0x00101010 — APPROVED** [review]
